### PR TITLE
Extract modelmesh part in helm chart

### PR DESCRIPTION
**Pull request by hhk7734**
**What this PR does / why we need it**:

- For serverless installation without modelmesh

**Which issue(s) this PR fixes** :

x

**Type of changes**

x

**Feature/Issue validation/testing**:

```shell
helm template kserve ./charts/kserve-resources \
  -n kserve \
  --set kserve.modelmesh.enabled=false \
  > kserve-helm.yaml
```

kserve-helm.yaml and [kserve.yaml](https://github.com/kserve/kserve/releases/download/v0.10.0/kserve.yaml)  should be approximately the same without CRDs.

**Special notes for your reviewer**:

x

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Add kserve.modelmesh.enabled option to kserve-resources helm chart
```

### CODE REVIEW

File: charts/kserve-resources/templates/clusterrole.yaml [hunk 1]
```diff
@@ -302,194 +304,4 @@ rules:
...
-      - watch
-  - apiGroups:
-      - serving.kserve.io
-    resources:
-      - clusterservingruntimes/status
-    verbs:
-      - get
-      - patch
-      - update
+  - update
```
Comment (yuzisun) [inline charts/kserve-resources/templates/clusterrole.yaml]: nit: newline

File: charts/kserve-resources/templates/clusterrolebinding.yaml [hunk 2]
```diff
@@ -22,21 +24,4 @@ roleRef:
...
-  name: modelmesh-controller-rolebinding
-roleRef:
-  apiGroup: rbac.authorization.k8s.io
-  kind: ClusterRole
-  name: modelmesh-controller-role
-subjects:
-- kind: ServiceAccount
-  name: modelmesh-controller
-  namespace: {{ .Release.Namespace }}
+  namespace: {{ .Release.Namespace }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/clusterrolebinding.yaml]: nit: newline

File: charts/kserve-resources/templates/clusterservingruntimes.yaml [hunk 3]
```diff
@@ -308,222 +316,4 @@ spec:
...
-          memory: 1Gi
-        limits:
-          cpu: "5"
-          memory: 1Gi
-  builtInAdapter:
-    serverType: torchserve
-    runtimeManagementPort: 7071
-    memBufferBytes: 134217728
-    modelLoadingTimeoutMillis: 90000
+          memory: 2Gi
```
Comment (yuzisun) [inline charts/kserve-resources/templates/clusterservingruntimes.yaml]: nit: newline

File: charts/kserve-resources/templates/deployment.yaml [hunk 4]
```diff
@@ -69,98 +69,4 @@ spec:
...
-          name: config-defaults
-          readOnly: true
-      serviceAccountName: modelmesh-controller
-      terminationGracePeriodSeconds: 10
-      volumes:
-      - configMap:
-          defaultMode: 420
-          name: model-serving-config-defaults
-        name: config-defaults
+          secretName: kserve-webhook-server-cert
```
Comment (yuzisun) [inline charts/kserve-resources/templates/deployment.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/clusterrole.yaml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      - watch
+  - apiGroups:
+      - serving.kserve.io
+    resources:
+      - clusterservingruntimes/status
+    verbs:
+      - get
+      - patch
+      - update
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/clusterrole.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/clusterrolebinding.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  name: modelmesh-controller-rolebinding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: modelmesh-controller-role
+subjects:
+- kind: ServiceAccount
+  name: modelmesh-controller
+  namespace: {{ .Release.Namespace }}
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/clusterrolebinding.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/clusterservingruntimes.yaml [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+          memory: 1Gi
+        limits:
+          cpu: "5"
+          memory: 1Gi
+  builtInAdapter:
+    serverType: torchserve
+    runtimeManagementPort: 7071
+    memBufferBytes: 134217728
+    modelLoadingTimeoutMillis: 90000
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/clusterservingruntimes.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/configmap.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        memory: "512Mi"
+    serviceAccountName: ""
+    metrics:
+      enabled: true
+    builtInServerTypes:
+      - triton
+      - mlserver
+      - ovms
+      - torchserve
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/configmap.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+          name: config-defaults
+          readOnly: true
+      serviceAccountName: modelmesh-controller
+      terminationGracePeriodSeconds: 10
+      volumes:
+      - configMap:
+          defaultMode: 420
+          name: model-serving-config-defaults
+        name: config-defaults
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/deployment.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/networkpolicy.yaml [hunk 10]
```diff
@@ -53,3 +56,4 @@ spec:
       app.kubernetes.io/managed-by: modelmesh-controller
   policyTypes:
   - Ingress
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/networkpolicy.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/role.yaml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+rules:
+- apiGroups:
+  - security.openshift.io
+  resourceNames:
+  - restricted
+  resources:
+  - securitycontextconstraints
+  verbs:
+  - use
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/role.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/rolebinding.yaml [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    app.kubernetes.io/name: modelmesh-controller
+  name: modelmesh-controller-restricted-scc
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: modelmesh-controller-restricted-scc
+subjects:
+- kind: ServiceAccount
+  name: modelmesh-controller
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/rolebinding.yaml]: nit: newline

File: charts/kserve-resources/templates/modelmesh/serviceaccount.yaml [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  labels:
+    app.kubernetes.io/instance: modelmesh-controller
+    app.kubernetes.io/managed-by: modelmesh-controller
+    app.kubernetes.io/name: modelmesh-controller
+  name: modelmesh-controller
+{{- end }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/modelmesh/serviceaccount.yaml]: nit: newline

File: charts/kserve-resources/templates/role.yaml [hunk 14]
```diff
@@ -38,59 +38,4 @@ rules:
...
-rules:
-- apiGroups:
-  - security.openshift.io
-  resourceNames:
-  - restricted
-  resources:
-  - securitycontextconstraints
-  verbs:
-  - use
+  - create
```
Comment (yuzisun) [inline charts/kserve-resources/templates/role.yaml]: nit: newline

File: charts/kserve-resources/templates/rolebinding.yaml [hunk 15]
```diff
@@ -9,36 +10,4 @@ roleRef:
...
-    app.kubernetes.io/name: modelmesh-controller
-  name: modelmesh-controller-restricted-scc
-roleRef:
-  apiGroup: rbac.authorization.k8s.io
-  kind: Role
-  name: modelmesh-controller-restricted-scc
-subjects:
-- kind: ServiceAccount
-  name: modelmesh-controller
+  namespace: {{ .Release.Namespace }}
```
Comment (yuzisun) [inline charts/kserve-resources/templates/rolebinding.yaml]: nit: newline

**Comment by yuzisun** [comment]
@hhk7734 Thanks for the PR and it looks good ! just a few nits for fixing the new lines

**Comment by sukumargaonkar** [comment]
should we make the modelmesh resources as a subchart?
that way we can enable or disable their creation via variables in values.yaml

Random example form google
Eg: https://stackoverflow.com/questions/54032974/helm-conditionally-install-subchart

File: charts/kserve-resources/templates/clusterrole.yaml (hunk 1 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/clusterrole.yaml]: ad539c4

File: charts/kserve-resources/templates/clusterrolebinding.yaml (hunk 2 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/clusterrolebinding.yaml]: ad539c4

File: charts/kserve-resources/templates/clusterservingruntimes.yaml (hunk 3 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/clusterservingruntimes.yaml]: ad539c4

File: charts/kserve-resources/templates/deployment.yaml (hunk 4 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/deployment.yaml]: ad539c4

File: charts/kserve-resources/templates/modelmesh/clusterrole.yaml (hunk 5 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/modelmesh/clusterrole.yaml]: ad539c4

File: charts/kserve-resources/templates/modelmesh/clusterrolebinding.yaml (hunk 6 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/modelmesh/clusterrolebinding.yaml]: ad539c4

File: charts/kserve-resources/templates/modelmesh/role.yaml (hunk 11 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/modelmesh/role.yaml]: ad539c4

File: charts/kserve-resources/templates/modelmesh/rolebinding.yaml (hunk 12 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/modelmesh/rolebinding.yaml]: ad539c4

File: charts/kserve-resources/templates/modelmesh/serviceaccount.yaml (hunk 13 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/modelmesh/serviceaccount.yaml]: ad539c4

File: charts/kserve-resources/templates/role.yaml (hunk 14 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/role.yaml]: ad539c4

File: charts/kserve-resources/templates/rolebinding.yaml (hunk 15 above)
Comment (hhk7734) [reply charts/kserve-resources/templates/rolebinding.yaml]: ad539c4

**Comment by hhk7734** [comment]
@yuzisun Please review again :)

**Comment by yuzisun** [comment]
> @yuzisun Please review again :)

@hhk7734 have you considered the sub chart idea @sukumargaonkar commented ? This avoids all the if checks

**Comment by hhk7734** [comment]
> > @yuzisun Please review again :)
> 
> @hhk7734 have you considered the sub chart idea @sukumargaonkar commented ? This avoids all the if checks

If this is turned into a subchart, version control for the modelmesh chart should also be considered. I am wondering if the kserve organization plans to manage this.

**Comment by yuzisun** [comment]
> > > @yuzisun Please review again :)
> > 
> > 
> > @hhk7734 have you considered the sub chart idea @sukumargaonkar commented ? This avoids all the if checks
> 
> If this is turned into a subchart, version control for the modelmesh chart should also be considered. I am wondering if the kserve organization plans to manage this.

@hhk7734 modelmesh is always released together with kserve, we do not have separate cadence for modelmesh.

**Comment by hhk7734** [comment]
@yuzisun If you want to apply subchart, I'll make the following changes. Do you want to use subchart?

CC. @sukumargaonkar 

---

### Directory layout

```shell
charts/
└── kserve-resources/
    ├── Chart.yaml
    ├── charts/
    │   └── modelmesh/
    │       ├── Chart.yaml
    │       └── templates/
    │           ├── clusterrole.yaml
    │           ├── clusterrolebinding.yaml
    │           ├── clusterservingruntimes.yaml
    │           ├── configmap.yaml
    │           ├── deployment.yaml
    │           ├── networkpolicy.yaml
    │           ├── role.yaml
    │           ├── rolebinding.yaml
    │           └── serviceaccount.yaml
    ├── templates/
    │   ├── certificate.yaml
    │   ├── clusterrole.yaml
    │   ├── clusterrolebinding.yaml
    │   ├── clusterservingruntimes.yaml
    │   ├── configmap.yaml
    │   ├── deployment.yaml
    │   ├── role.yaml
    │   ├── rolebinding.yaml
    │   ├── service.yaml
    │   ├── serviceaccount.yaml
    │   └── webhookconfiguration.yaml
    └── values.yaml
```

---

### Add modelmesh `Chart.yaml`

```yaml
apiVersion: v1
name: modelmesh
version: v0.10.0
description: Helm chart for deploying kserve resources
keywords:
  - kserve
  - modelmesh
sources:
  - http://github.com/kserve/kserve
```

### Add modelmesh dependency to kserve-resources `Chart.yaml`

```yaml
...
dependencies:
  - name: modelmesh
    version: v0.10.0
    condition: modelmesh.enabled
```

---

### Change the hierarchy of kserve-resources `values.yaml`

```yaml
kserve:
  ...
  modelmesh:
    enabled: true
    ...
  ...
```

to

```yaml
kserve:
  ...
modelmesh:
  enabled: true
  ...
```

---

### Change the modemesh values


```yaml
        image: {{ .Values.kserve.modelmesh.controller.image }}:{{ .Values.kserve.modelmesh.controller.tag }}
```

to

```yaml
        image: {{ .Values.controller.image }}:{{ .Values.controller.tag }}
```

**Comment by hhk7734** [comment]
@yuzisun

How about reflecting this PR without modification first for backwards compatibility, and later if a subchart is needed?

**Comment by hhk7734** [comment]
@yuzisun

Reflected changes in the master chart.

**Comment by yuzisun** [comment]
Thanks @hhk7734 !

/lgtm
/approve

**Pull request merged**